pkg/services/authz/server.go

```diff
@@ -70,7 +82,120 @@ func (l *legacyServer) AuthorizeFuncOverride(ctx context.Context) error {
 }
 
-func (l *legacyServer) Check(context.Context, *authzv1.CheckRequest) (*authzv1.CheckResponse, error) {
-	// FIXME: implement for legacy access control
-	return nil, errors.New("unimplemented")
+func wrapErr(err error) error {
+	return status.Error(codes.Internal, fmt.Errorf("authz check failed: %w", err).Error())
+}
+
+func validateRequest(req *authzv1.CheckRequest) error {
+	if req.GetGroup() == "" {
+		return status.Error(codes.InvalidArgument, "group is required")
+	}
+	if req.GetResource() == "" {
+		return status.Error(codes.InvalidArgument, "resource is required")
+	}
+	if req.GetVerb() == "" {
+		return status.Error(codes.InvalidArgument, "verb is required")
+	}
+	if req.GetSubject() == "" {
+		return status.Error(codes.InvalidArgument, "subject is required")
+	}
+	return nil
+}
+
+func (l *legacyServer) Check(ctx context.Context, req *authzv1.CheckRequest) (*authzv1.CheckResponse, error) {
+	ctx, span := l.tracer.Start(ctx, "authz.Check")
+	defer span.End()
+	ctxLogger := l.logger.FromContext(ctx)
+
+	deny := &authzv1.CheckResponse{Allowed: false}
+	if err := validateRequest(req); err != nil {
+		ctxLogger.Error("invalid request", "error", err)
+		return deny, err
+	}
+
+	namespace := req.GetNamespace()
+	info, err := claims.ParseNamespace(namespace)
+	// FIXME: Remove the stackID check once the ParseNamespace function does not return orgID 1 for stacks namespaces
+	if err != nil || info.OrgID == 0 || info.StackID != 0 {
+		ctxLogger.Error("invalid namespace", "namespace", namespace, "error", err)
+		return deny, status.Error(codes.InvalidArgument, "invalid namespace: "+namespace)
+	}
+
+	// Get the RBAC action associated with the request
+	action, ok := l.mapper.Action(req.Group, req.Resource, req.Verb)
+	if !ok {
+		ctxLogger.Error("could not find associated rbac action", "group", req.Group, "resource", req.Resource, "verb", req.Verb)
+		return deny, wrapErr(fmt.Errorf("could not find associated rbac action"))
+	}
+
+	// Get the user from the subject
+	user, err := l.authnSvc.ResolveIdentity(ctx, info.OrgID, req.Subject)
+	if err != nil {
+		// TODO: should probably distinguish between not found and other errors
+		ctxLogger.Error("could not resolve identity", "subject", req.Subject, "orgId", info.OrgID)
+		return deny, wrapErr(fmt.Errorf("could not resolve identity"))
+	}
+
+	// Check if the user has the action solely
+	if req.Name == "" && req.Folder == "" {
+		ev := accesscontrol.EvalPermission(action)
+		hasAccess, err := l.ac.Evaluate(ctx, user, ev)
+		if err != nil {
+			ctxLogger.Error("could not evaluate permission", "subject", req.Subject, "orgId", info.OrgID, "action", action)
+			return deny, wrapErr(fmt.Errorf("could not evaluate permission"))
+		}
+
+		return &authzv1.CheckResponse{Allowed: hasAccess}, nil
+	}
+
+	scopes := make([]string, 0, 1)
+	// If a parent is specified: Check if the user has access to any of the parent folders
+	if req.Folder != "" {
+		scopes, err = l.getFolderTree(ctx, info.OrgID, req.Folder)
```

Comment:
That's a bit tricky moment: inside the Evaluate() function we perform scope mutation and if we have something like `dashboards:read` with scope `dashboard:uid:1`, then evaluator will call resolvers for the specific type (dashboard) and scope will be mutated to set of scopes with all the parent folders. Do we really need to build parents explicitly?

Reply:
Oh good point! if the mutators are already handling the folder tree, then we are good right?
Alternatively I can use the access control `WithoutResolvers`:
https://github.com/grafana/grafana/blob/main/pkg/services/accesscontrol/accesscontrol.go#L33

Reply:
Yes, might be good. I like more explicit approach you have with `getFolderTree()`, just want to avoid double calls.

Reply:
ok change done, thanks for this :100: